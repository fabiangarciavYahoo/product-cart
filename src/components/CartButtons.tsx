import React, { FC, useCallback, useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';

export interface CartButtonsProps {
  className?: string;
}

const CartButtons: FC<CartButtonsProps> = ({ className }) => {
  const { counter, increaseBy, maxCount } = useContext(ProductContext);

  const isMaxCountReached = useCallback(
    () => !!maxCount && counter === maxCount,
    [counter, maxCount]
  );

  return (
    <div className={`${styles.buttonsContainer} ${className}`}>
      <button
        className={styles.buttonMinus}
        onClick={() => counter !== 0 && increaseBy(-1)}
      >
        -
      </button>
      <div className={styles.countLabel}>{counter}</div>
      <button
        className={`${styles.buttonAdd} ${
          isMaxCountReached() ? styles.buttonDisabled : ''
        }`}
        onClick={() => increaseBy(1)}
      >
        +
      </button>
    </div>
  );
};

export default CartButtons;
