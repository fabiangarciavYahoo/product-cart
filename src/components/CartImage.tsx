import React, { FC, useContext } from 'react';
import useProduct from '../hooks/useProduct';
import styles from '../styles/styles.module.css';
import { ProductContext } from './ProductCard';

export interface CartImageProps {
  className?: string;
}

export const CartImage: FC<CartImageProps> = ({ className }) => {
  const {
    product: { name },
    product,
  } = useContext(ProductContext);
  const { imagePath, onImageError } = useProduct({ product });
  return (
    <img
      className={`${styles.productImg} ${className}`}
      src={imagePath}
      alt={name}
      onError={onImageError}
    />
  );
};

export default CartImage;
