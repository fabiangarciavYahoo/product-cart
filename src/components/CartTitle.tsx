import React, { FC, useContext } from 'react';
import styles from '../styles/styles.module.css';
import { ProductContext } from './ProductCard';

export interface CartTitleProps {
  className?: string;
  style?: React.CSSProperties;
}

export const CartTitle: FC<CartTitleProps> = ({ style, className }) => {
  const {
    product: { name, price },
  } = useContext(ProductContext);
  return (
    <span style={style} className={`${styles.productDescription} ${className}`}>
      {name} $ {price}
    </span>
  );
};

export default CartTitle;
