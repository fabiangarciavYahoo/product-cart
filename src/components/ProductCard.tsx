import styles from '../styles/styles.module.css';
import { IProductContextProps, ProductCardComponent } from '../interfaces';
import React, { createContext } from 'react';
import CartImage from './CartImage';
import CartButtons from './CartButtons';
import CartTitle from './CartTitle';
import useProduct from '../hooks/useProduct';

export const ProductContext = createContext({} as IProductContextProps);
const { Provider } = ProductContext;

const ProductCard: ProductCardComponent = ({
  children,
  product,
  className = '',
  style = {},
  onChange,
  value,
  initialValues,
}) => {
  const {
    counter,
    increaseBy,
    maxCount,
    reset,
    isMaxCountReached,
  } = useProduct({
    product,
    onChange,
    value,
    initialValues,
  });

  return (
    <Provider
      value={{
        counter,
        increaseBy,
        product,
        maxCount,
      }}
    >
      <div style={style} className={`${styles.productCard} ${className}`}>
        {children({
          count: counter,
          isMaxCountReached: isMaxCountReached(),
          maxCount,
          product,
          increaseBy,
          reset,
        })}
      </div>
    </Provider>
  );
};

ProductCard.Image = CartImage;
ProductCard.Title = CartTitle;
ProductCard.Buttons = CartButtons;

export default ProductCard;
