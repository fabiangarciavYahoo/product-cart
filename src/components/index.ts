import ProductCartHOC from './ProductCard';
import CartButtons from './CartButtons';
import CartImage from './CartImage';
import CartTitle from './CartTitle';

export * from './CartImage';
export * from './CartTitle';
export * from './CartButtons';

export const ProductCard = Object.assign(ProductCartHOC, {
  Image: CartImage,
  Title: CartTitle,
  Buttons: CartButtons,
});
export default ProductCard;
