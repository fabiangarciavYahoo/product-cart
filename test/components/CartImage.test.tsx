import React from 'react';
import { render } from '@testing-library/react';
import CartImage from '../../src/components/CartImage';
import { ProductContext } from '../../src/components/ProductCard';

describe('CartImage', () => {
  test('renders without crashing with context', () => {
    const mockProduct = {
      id: 1,
      name: 'Test Product',
      price: 100,
      image: 'test-image.jpg',
    };
    const mockContextValue = {
      product: mockProduct,
      counter: 1,
      increaseBy: jest.fn(),
    };
    const { asFragment } = render(
      <ProductContext.Provider value={mockContextValue}>
        <CartImage className="poppet" />
      </ProductContext.Provider>
    );
    const plainText = asFragment().textContent;
    console.log('Plain text content:', plainText);
    expect(asFragment()).toMatchSnapshot();
  });
});
