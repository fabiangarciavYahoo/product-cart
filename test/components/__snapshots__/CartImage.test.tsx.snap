// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartImage renders without crashing with context 1`] = `
<DocumentFragment>
  <img
    alt="Test Product"
    class="productImg poppet"
    src="test-image.jpg"
  />
</DocumentFragment>
`;
